File: src/spritz/surfaces/triangle.py
import numpy as np

from .surface import Surface
from ..raytracing import Ray, Intersection

class Triangle(Surface):
    """Spritz Triangle"""

    def __init__(self, a, b, c, material):
        """Triangle with vertices a, b, c in clock-wise orientation.

        Args:
            a (ArrayLike): 1st vertex
            b (ArrayLike): 2nd vertex
            c (ArrayLike): 3rd vertex
        """
        self.v1 = np.array(a, dtype=float)
        self.v2 = np.array(b, dtype=float)
        self.v3 = np.array(c, dtype=float)
        self.material = material

    def __repr__(self):
        v1 = f"({self.v1[0]}, {self.v1[1]}, {self.v1[2]})"
        v2 = f"({self.v2[0]}, {self.v2[1]}, {self.v2[2]})"
        v3 = f"({self.v3[0]}, {self.v3[1]}, {self.v3[2]})"
        return f"<Triangle at {v1}->{v2}->{v3}>"

    def hit(self, ray, t0=0, t1=np.inf):
        """An intersection of a ray with the triangle occurs at time t if:

        `ray_origin` + `t`*`ray_direction` = `v1` + `\beta(v2 - v1)` + `gamma(v3 - 1)`

        This is a linear system solvable with Cramer's rule.
        
        Args:
            ray (Ray): Ray to check
            t0 (float): Start of time interval
            t1 (float): End of time interval

        Returns: None if no hit, otherwise Intersection object
        """
        # Build linear system:
        # [a d g] [\beta ]    [j]
        # |b e h| |\gamma| =  |k|
        # [c f i] [  t   ]    [l]
        v2v1 = self.v1 - self.v2 # Edge v2 -> v1
        v3v1 = self.v1 - self.v3 # Edge v3 -> v1
        a, b, c = v2v1 
        d, e, f = v3v1
        g, h, i = ray.direction
        j, k, l = self.v1 - ray.origin # Vector from ray origin to v1

        # Solve with Cramer's rule
        ei_hf = e*i - h*f
        gf_di = g*f - d*i
        dh_eg = d*h - e*g
        det = a*ei_hf + b*gf_di + c*dh_eg
        if det == 0:
            print("M fails")
            return None
        
        ak_jb = a*k - j*b
        jc_al = j*c - a*l
        bl_kc = b*l - k*c

        # Early stopping conditions
        t = -(f*ak_jb + e*jc_al + d*bl_kc) / det
        if t0 <= t <= t1:
            gamma = (i*ak_jb + h*jc_al + g*bl_kc) / det
            if 0 <= gamma <= 1:
                beta = (j*ei_hf + k*gf_di + l*dh_eg) / det
                if 0 <= beta <= 1 - gamma:
                    normal = np.cross(v2v1, v3v1)
                    normal /= np.linalg.norm(normal)
                    return Intersection(self, t, normal)
        return None

                
        

File: src/surfaces/triangle.cpp
#include <cmath>

#include "surfaces.h"

#define min(a,b) (((a) < (b)) ? (a) : (b))

// ===== CONSTRUCTORS =====
Triangle::Triangle(Vec3 vert1, Vec3 vert2, Vec3 vert3, Material* mat) {
    v1 = vert1;
    v2 = vert2;
    v3 = vert3;
    edge1 = v2 - v1;
    edge2 = v3 - v1;
    normal = edge1.cross(edge2).normalized();
    material = mat;
}

// ===== METHODS =====

// An intersection of a ray with the triangle occurs at time t if:
// `ray_origin` + `t`*`ray_direction` = `v1` + `\beta(v2 - v1)` + `gamma(v3 - 1)`
// This is a linear system solvable with Cramer's rule.
Intersection Triangle::hit(const Ray& ray, double t0, double t1) {
    
    Intersection hit; // empty hit record denotes a miss

    const double EPS = 1e-8;
    Vec3 pvec = ray.d.cross(edge2);
    double det = edge1.dot(pvec);

    if (fabs(det) < EPS) return hit; // Ray parallel to triangle

    double invDet = 1.0 / det;
    Vec3 tvec = ray.o - v1;
    double u = tvec.dot(pvec) * invDet;
    if (u < 0.0 || u > 1.0) return hit;

    Vec3 qvec = tvec.cross(edge1);
    double v = ray.d.dot(qvec) * invDet;
    if (v < 0.0 || u + v > 1.0) return hit;

    double t = edge2.dot(qvec) * invDet;
    if (t < t0 || t > t1) return hit;

    hit.t = t;
    if (normal.dot(ray.d) > 0) {
        hit.normal = normal * -1;
    }
    else {
        hit.normal = normal;
    }
    hit.surface = this;
    return hit;
}
